### PR TITLE
Improvements to API for large file support

### PR DESCRIPTION
**Pull request by nmoinvaz**
See #445 for background.

For gzseek, gzoffset, gzopen, adler32_combine, crc32_combine and crc32_combine_gen:

- Export 32-bit and 64-bit versions for zlib-compatible api
- Export only 64-bit version (without 64 suffix) for zlib-ng native api.

### CODE REVIEW

**Comment by nmoinvaz** [comment]
Zlib-compatible API exports:
```
1 (0x0001), N/A, adler32, 0x00001000, None
2 (0x0002), 1 (0x00000001), adler32_combine, 0x00001210, None
3 (0x0003), 2 (0x00000002), adler32_combine64, 0x00001230, None
4 (0x0004), 3 (0x00000003), adler32_z, 0x00001000, None
5 (0x0005), 4 (0x00000004), compress, 0x000015b0, None
6 (0x0006), 5 (0x00000005), compress2, 0x000015d0, None
7 (0x0007), 6 (0x00000006), compressBound, 0x000016a0, None
8 (0x0008), 7 (0x00000007), crc32, 0x00001730, None
9 (0x0009), 8 (0x00000008), crc32_combine, 0x00001770, None
10 (0x000a), 9 (0x00000009), crc32_combine64, 0x00001790, None
11 (0x000b), 10 (0x0000000a), crc32_combine_gen, 0x00001830, None
12 (0x000c), 11 (0x0000000b), crc32_combine_gen64, 0x00001850, None
13 (0x000d), 12 (0x0000000c), crc32_combine_op, 0x00001980, None
14 (0x000e), 13 (0x0000000d), crc32_z, 0x00001730, None
15 (0x000f), 14 (0x0000000e), deflate, 0x00001f60, None
16 (0x0010), 15 (0x0000000f), deflateBound, 0x000028d0, None
17 (0x0011), 16 (0x00000010), deflateCopy, 0x00002a10, None
18 (0x0012), 17 (0x00000011), deflateEnd, 0x00002c10, None
19 (0x0013), 18 (0x00000012), deflateGetDictionary, 0x00002d10, None
20 (0x0014), 19 (0x00000013), deflateInit2_, 0x00002dc0, None
21 (0x0015), 20 (0x00000014), deflateInit_, 0x00003050, None
22 (0x0016), 21 (0x00000015), deflateParams, 0x00003080, None
23 (0x0017), 22 (0x00000016), deflatePending, 0x00003240, None
24 (0x0018), 23 (0x00000017), deflatePrime, 0x000032c0, None
25 (0x0019), 24 (0x00000018), deflateReset, 0x000033b0, None
26 (0x001a), 25 (0x00000019), deflateResetKeep, 0x00003490, None
27 (0x001b), 26 (0x0000001a), deflateSetDictionary, 0x00003590, None
28 (0x001c), 27 (0x0000001b), deflateSetHeader, 0x000037b0, None
29 (0x001d), 28 (0x0000001c), deflateTune, 0x00003820, None
30 (0x001e), 29 (0x0000001d), get_crc_table, 0x00001ae0, None
31 (0x001f), 30 (0x0000001e), gzbuffer, 0x0000cfd0, None
32 (0x0020), 31 (0x0000001f), gzclearerr, 0x0000d020, None
33 (0x0021), 32 (0x00000020), gzclose, 0x0000cab0, None
34 (0x0022), 33 (0x00000021), gzclose_r, 0x0000dc40, None
35 (0x0023), 34 (0x00000022), gzclose_w, 0x0000e5a0, None
36 (0x0024), 35 (0x00000023), gzdirect, 0x0000dcc0, None
37 (0x0025), 36 (0x00000024), gzdopen, 0x0000d080, None
38 (0x0026), 37 (0x00000025), gzeof, 0x0000d0d0, None
39 (0x0027), 38 (0x00000026), gzerror, 0x0000d0f0, None
40 (0x0028), 39 (0x00000027), gzflush, 0x0000e660, None
41 (0x0029), 40 (0x00000028), gzfread, 0x0000dd00, None
42 (0x002a), 41 (0x00000029), gzfwrite, 0x0000e6c0, None
43 (0x002b), 42 (0x0000002a), gzgetc, 0x0000dd80, None
44 (0x002c), 43 (0x0000002b), gzgetc_, 0x0000dd80, None
45 (0x002d), 44 (0x0000002c), gzgets, 0x0000dde0, None
46 (0x002e), 45 (0x0000002d), gzoffset, 0x0000d140, None
47 (0x002f), 46 (0x0000002e), gzoffset64, 0x0000d1c0, None
48 (0x0030), 47 (0x0000002f), gzopen, 0x0000d220, None
49 (0x0031), 48 (0x00000030), gzopen64, 0x0000d220, None
50 (0x0032), 49 (0x00000031), gzopen_w, 0x0000d240, None
51 (0x0033), 50 (0x00000032), gzprintf, 0x0000e730, None
52 (0x0034), 51 (0x00000033), gzputc, 0x0000e750, None
53 (0x0035), 52 (0x00000034), gzputs, 0x0000e810, None
54 (0x0036), 53 (0x00000035), gzread, 0x0000df00, None
55 (0x0037), 54 (0x00000036), gzrewind, 0x0000d260, None
56 (0x0038), 55 (0x00000037), gzseek, 0x0000d2b0, None
57 (0x0039), 56 (0x00000038), gzsetparams, 0x0000e880, None
58 (0x003a), 57 (0x00000039), gztell, 0x0000d4d0, None
59 (0x003b), 58 (0x0000003a), gztell64, 0x0000d550, None
60 (0x003c), 59 (0x0000003b), gzungetc, 0x0000df60, None
61 (0x003d), 60 (0x0000003c), gzvprintf, 0x0000e920, None
62 (0x003e), 61 (0x0000003d), gzwrite, 0x0000ea50, None
63 (0x003f), 62 (0x0000003e), inflate, 0x00005560, None
64 (0x0040), 63 (0x0000003f), inflateBack, 0x00007d20, None
65 (0x0041), 64 (0x00000040), inflateBackEnd, 0x00008b00, None
66 (0x0042), 65 (0x00000041), inflateBackInit_, 0x00008b40, None
67 (0x0043), 66 (0x00000042), inflateCodesUsed, 0x00007110, None
68 (0x0044), 67 (0x00000043), inflateCopy, 0x00007140, None
69 (0x0045), 68 (0x00000044), inflateEnd, 0x000072c0, None
70 (0x0046), 69 (0x00000045), inflateGetDictionary, 0x00007330, None
71 (0x0047), 70 (0x00000046), inflateGetHeader, 0x000073b0, None
72 (0x0048), 71 (0x00000047), inflateInit2_, 0x00007400, None
73 (0x0049), 72 (0x00000048), inflateInit_, 0x000074d0, None
74 (0x004a), 73 (0x00000049), inflateMark, 0x000074f0, None
75 (0x004b), 74 (0x0000004a), inflatePrime, 0x00007570, None
76 (0x004c), 75 (0x0000004b), inflateReset, 0x000075f0, None
77 (0x004d), 76 (0x0000004c), inflateReset2, 0x00007650, None
78 (0x004e), 77 (0x0000004d), inflateResetKeep, 0x00007750, None
79 (0x004f), 78 (0x0000004e), inflateSetDictionary, 0x00007850, None
80 (0x0050), 79 (0x0000004f), inflateSync, 0x00007930, None
81 (0x0051), 80 (0x00000050), inflateSyncPoint, 0x00007ab0, None
82 (0x0052), 81 (0x00000051), inflateUndermine, 0x00007b00, None
83 (0x0053), 82 (0x00000052), inflateValidate, 0x00007b50, None
84 (0x0054), 83 (0x00000053), uncompress, 0x0000b410, None
85 (0x0055), 84 (0x00000054), uncompress2, 0x0000b430, None
86 (0x0056), 85 (0x00000055), zError, 0x0000b550, None
87 (0x0057), 86 (0x00000056), zlibCompileFlags, 0x0000b570, None
88 (0x0058), 87 (0x00000057), zlibVersion, 0x0000b580, None
```

Zlib-ng native API exports:
```
1 (0x0001), N/A, zlibng_version, 0x0000b7e0, None
2 (0x0002), 1 (0x00000001), zng_adler32, 0x000011f0, None
3 (0x0003), 2 (0x00000002), zng_adler32_combine, 0x00001210, None
4 (0x0004), 3 (0x00000003), zng_adler32_z, 0x000011f0, None
5 (0x0005), 4 (0x00000004), zng_compress, 0x00001570, None
6 (0x0006), 5 (0x00000005), zng_compress2, 0x00001590, None
7 (0x0007), 6 (0x00000006), zng_compressBound, 0x00001660, None
8 (0x0008), 7 (0x00000007), zng_crc32, 0x00001820, None
9 (0x0009), 8 (0x00000008), zng_crc32_combine, 0x00001860, None
10 (0x000a), 9 (0x00000009), zng_crc32_combine_gen, 0x000018e0, None
11 (0x000b), 10 (0x0000000a), zng_crc32_combine_op, 0x000019f0, None
12 (0x000c), 11 (0x0000000b), zng_crc32_z, 0x00001820, None
13 (0x000d), 12 (0x0000000c), zng_deflate, 0x00002830, None
14 (0x000e), 13 (0x0000000d), zng_deflateBound, 0x000031a0, None
15 (0x000f), 14 (0x0000000e), zng_deflateCopy, 0x000032e0, None
16 (0x0010), 15 (0x0000000f), zng_deflateEnd, 0x000034e0, None
17 (0x0011), 16 (0x00000010), zng_deflateGetDictionary, 0x000035e0, None
18 (0x0012), 17 (0x00000011), zng_deflateGetParams, 0x00003690, None
19 (0x0013), 18 (0x00000012), zng_deflateInit2_, 0x000037d0, None
20 (0x0014), 19 (0x00000013), zng_deflateInit_, 0x00003a60, None
21 (0x0015), 20 (0x00000014), zng_deflateParams, 0x00003a90, None
22 (0x0016), 21 (0x00000015), zng_deflatePending, 0x00003c50, None
23 (0x0017), 22 (0x00000016), zng_deflatePrime, 0x00003cd0, None
24 (0x0018), 23 (0x00000017), zng_deflateReset, 0x00003dc0, None
25 (0x0019), 24 (0x00000018), zng_deflateResetKeep, 0x00003ea0, None
26 (0x001a), 25 (0x00000019), zng_deflateSetDictionary, 0x00003fa0, None
27 (0x001b), 26 (0x0000001a), zng_deflateSetHeader, 0x000041c0, None
28 (0x001c), 27 (0x0000001b), zng_deflateSetParams, 0x00004230, None
29 (0x001d), 28 (0x0000001c), zng_deflateTune, 0x00004430, None
30 (0x001e), 29 (0x0000001d), zng_get_crc_table, 0x00001a20, None
31 (0x001f), 30 (0x0000001e), zng_inflate, 0x00005960, None
32 (0x0020), 31 (0x0000001f), zng_inflateBack, 0x00007fb0, None
33 (0x0021), 32 (0x00000020), zng_inflateBackEnd, 0x00008d90, None
34 (0x0022), 33 (0x00000021), zng_inflateBackInit_, 0x00008dd0, None
35 (0x0023), 34 (0x00000022), zng_inflateCodesUsed, 0x00007510, None
36 (0x0024), 35 (0x00000023), zng_inflateCopy, 0x00007540, None
37 (0x0025), 36 (0x00000024), zng_inflateEnd, 0x000076c0, None
38 (0x0026), 37 (0x00000025), zng_inflateGetDictionary, 0x00007730, None
39 (0x0027), 38 (0x00000026), zng_inflateGetHeader, 0x000077b0, None
40 (0x0028), 39 (0x00000027), zng_inflateInit2_, 0x00007800, None
41 (0x0029), 40 (0x00000028), zng_inflateInit_, 0x000078d0, None
42 (0x002a), 41 (0x00000029), zng_inflateMark, 0x000078f0, None
43 (0x002b), 42 (0x0000002a), zng_inflatePrime, 0x00007970, None
44 (0x002c), 43 (0x0000002b), zng_inflateReset, 0x000079f0, None
45 (0x002d), 44 (0x0000002c), zng_inflateReset2, 0x00007a50, None
46 (0x002e), 45 (0x0000002d), zng_inflateResetKeep, 0x00007b50, None
47 (0x002f), 46 (0x0000002e), zng_inflateSetDictionary, 0x00007c50, None
48 (0x0030), 47 (0x0000002f), zng_inflateSync, 0x00007d30, None
49 (0x0031), 48 (0x00000030), zng_inflateSyncPoint, 0x00007eb0, None
50 (0x0032), 49 (0x00000031), zng_inflateUndermine, 0x00007f00, None
51 (0x0033), 50 (0x00000032), zng_inflateValidate, 0x00007f50, None
52 (0x0034), 51 (0x00000033), zng_uncompress, 0x0000b6a0, None
53 (0x0035), 52 (0x00000034), zng_uncompress2, 0x0000b6c0, None
54 (0x0036), 53 (0x00000035), zng_zError, 0x0000b830, None
55 (0x0037), 54 (0x00000036), zng_zlibCompileFlags, 0x0000b850, None
```

File: README.md [hunk 1]
```diff
@@ -1,6 +1,6 @@
 ## zlib-ng
 *zlib data compression library for the next generation systems*
-
+test
```
Comment (mtl1979) [inline README.md]: test?

**Comment by mtl1979** [comment]
```
zlib-ng.def : error LNK2001: unresolved external symbol zng_adler32_combine64
zlib-ng.def : error LNK2001: unresolved external symbol zng_crc32_combine64
```

File: README.md (hunk 1 above)
Comment (nmoinvaz) [reply README.md]: Fixed.

**Comment by nmoinvaz** [comment]
That is fixed. Now some CI failing on `example64` but I have no idea why.

**Comment by mtl1979** [comment]
It fails on compat mode due to some issue with gzread... As example.c hasn't been modified, it is most likely something that was changed in zlib.h or command line for compiling example.c

**Comment by nmoinvaz** [comment]
Btw, win32/zlib.def is missing these entries, is that expected?

image

**Comment by mtl1979** [comment]
It was originally meant to be compiled with gz* functions disabled. Each flag combination needs own exports definition file, so it was limited to 2 combinations only.

**Comment by nmoinvaz** [comment]
Weird off by one error when z_off_t == z_off64_t.

```
zoff 8
zoff64 8
read 27 (! should be 28)
gzread err: 
zlib version 1.2.11.zlib-ng = 0x12bf, compile flags = 0x55
```

File: test/example.c [hunk 2]
```diff
@@ -138,7 +142,11 @@ void test_gzio(const char *fname, unsigned char *uncompr, z_size_t uncomprLen) {
     }
     /* Read uncompressed data - hello, hello! string twice */
     strcpy((char*)uncompr, "garbages");
-    if (PREFIX(gzread)(file, uncompr, (unsigned)uncomprLen) != (int)(len + len)) {
+    read = PREFIX(gzread)(file, uncompr, (unsigned)uncomprLen);
+    fprintf(stderr, "read %d\n", read);
+    if (read != (int)(len + len)) {
+        read =  PREFIX(gzread)(file, uncompr, (unsigned)uncomprLen);
```
Comment (mtl1979) [inline test/example.c]: Should print contents of "uncompr" as hexadecimal before attempting to read more bytes...

**Comment by Dead2** [comment]
This is looking very good!
Might I suggest making minigzip 64bit too, so we no longer have to build both minigzip and minigzip64? Doesn't matter whether we build compat or native.

**Comment by nmoinvaz** [comment]
The problem is setting `_FILE_OFFSET_BITS=64`. It changes `off_t` and subsequently `z_off_t` to 64-bit. If zlib is compiled with `_FILE_OFFSET_BITS=32` which is what is happening, and example64 is compiled with `_FILE_OFFSET_BITS=64` then the function prototypes will be wrong in example64 when including _zlib.h_.

One option is to just set `#define z_off_t unsigned int` instead of `#define z_off_t off_t`.

**Comment by Dead2** [comment]
But why compile with FILE_OFFSET_BITS=32 at all?
I am not talking about minigzipsh here, but the staticly compiled minigzip and minigzip64.
I mean we are compiling them both already, but why keep both? Does minigzip add any value over what minigzip64 already does?

**Comment by nmoinvaz** [comment]
We are not compiling with `_FILE_OFFSET_BITS=32` but I just meant to say that it is set to `64` in _example64_ and NOT being set in _zlibstatic_ causing the width of `z_off_t` to be different in each case. This is important for when `z_off_t` is used in zlib.h header function prototypes.

We can nuke the `example64` and `minigzip64` but they do highlight an issue - that our definition of `z_off_t` is tied to `off_t`. example64 and zlib would have to be compiled with the same `_FILE_OFFSET_BITS` value and that is currently not the case. That is why I suggest having `z_off_t = unsigned int`.

Another alternative would be to just not have `example64` or `minigzip64` and don't care about `_FILE_OFFSET_BITS`. But then if somebody did decide to compile with `_FILE_OFFSET_BITS=64` it would have to be set for both projects. If somebody tried to build a shared library with `_FILE_OFFSET_BITS=64` the function definitions would be off for 32-bit versions of _gzseek, gzoffset, gztell_ because it would be using 64-bit _z_off_t_ instead of 32-bit _z_off_t_.

**Comment by nmoinvaz** [comment]
After reading more about `_FILE_OFFSET_BITS`:

> ...Whereas _LARGEFILE64_SOURCE makes the 64 bit interface available as an additional interface, _FILE_OFFSET_BITS allows the 64 bit interface to replace the old interface. 

> If _FILE_OFFSET_BITS is undefined, or if it is defined to the value 32, nothing changes. The 32 bit interface is used and types like off_t have a size of 32 bits on 32 bit systems.

> If the macro is defined to the value 64, the large file interface replaces the old interface. I.e., the functions are not made available under different names (as they are with _LARGEFILE64_SOURCE). Instead the old function names now reference the new functions, e.g., a call to fseeko now indeed calls fseeko64.

> ...On 64 bit systems this macro has no effect since the *64 functions are identical to the normal functions.

https://www.gnu.org/software/libc/manual/html_node/Feature-Test-Macros.html

I think the best thing to do would be to remove the `example64` and `minigzip64` targets. Anybody who is compiling with `_FILE_OFFSET_BITS=64` will have to deal with the ramifications of that themselves.

**Comment by nmoinvaz** [comment]
@Dead2 this is ready for your review. Let me know if you want something different.

**Comment by Dead2** [comment]
Won't the removal of "_FILE_OFFSET_BITS=64" make 32bit minigzip unable to compress files bigger than 2GiB?

You said above:
" it is set to 64 in example64 and NOT being set in zlibstatic causing the width of z_off_t to be different in each case."
What I am failing to understand about the above statement is why minigzip64 is working currently? Is this not the current state of things? Or is minigzip64 currently broken?

Also, if off_t is the problem, why not change to off64_t in minigzip directly, and use the 64bit zlib-ng functions exclusively?


Other than the above, this PR looks very good.

**Comment by nmoinvaz** [comment]
> Won't the removal of "_FILE_OFFSET_BITS=64" make 32bit minigzip unable to compress files bigger than 2GiB?

It is my understanding that `_FILE_OFFSET_BITS` only controls the width of `off_t` and which interface functions are available under which name. 64-bit file access is still available under 64-bit function names (ie: fseek64 etc).

It is currently working properly with example64 and minigzip64 with the zlib-ng API because the zlib-ng.h header is redirecting the 32-bit functions (gzseek, gztell, gzoffset) to 64-bit versions (gzseek64, gztell64, gzoffset) when `_FILE_OFFSET_BITS=64`, so the 32-bit versions would never be called. But if the 32-bit versions were somehow called directly (say through shared library/dlopen method) then there would have been a problem.

In this PR, we do away with the 32-bit functions for zlib-ng API and use `z_off64_t` in all functions.

Also this PR does away with caring about `_FILE_OFFSET_BITS` in zlib API. For example gseek will always be 32-bit and gzseek64 will always be 64-bit. What do you think? Should we add back zlib API compatibility so if `_FILE_OFFSET_BITS=64` then gzseek will point to gzseek64?

**Comment by Dead2** [comment]
Yes, I think we probably should add that back in the compat api, otherwise it is not really compatible.

I also think that minigzip should explicitly target the 64bit parts of the api, even when compiled using the compat api without `_FILE_OFFSET_BITS=64`.

**Comment by nmoinvaz** [comment]
I have added back the `_FILE_OFFSET_BITS` functionality back to zlib.h.

**Comment by nmoinvaz** [comment]
Rebased

**Pull request merged**